modules/ipp/scanner.go

```diff
@@ -566,49 +551,60 @@ func (scanner *Scanner) newIPPScan(target *zgrab2.ScanTarget) *scan {
 	return &newScan
 }
 
+// TODO: Do you want to retry with TLS for all versions? Just one's you've already tried? Haven't tried? Just the same version?
+func (scanner *Scanner) tryGrabForVersions(target *zgrab2.ScanTarget, versions *[]version) (*scan, *zgrab2.ScanError) {
+	scan := scanner.newIPPScan(target)
+	// TODO: Implement scan.Cleanup()
+	var err *zgrab2.ScanError
+	for i := 0; i < len(*versions); i++ {
+		err = scanner.Grab(scan, target, &(*versions)[i])
+		if err != nil && err.Err == ErrVersionNotSupported && i < len(*versions)-1 {
+			continue
+		}
+		break
+	}
+	return scan, err
+}
+
+// TODO: Incorporate status into this? I don't think so, b/c with certain statuses, we should return
+// early, so special casing seems to make sense
+func (scan *scan) shouldReportResult(scanner *Scanner) bool {
+	if scan.results.Response != nil {
+		return true
+	} else if scanner.config.IPPSecure {
+		l := scan.results.TLSLog
+		return l != nil && l.HandshakeLog != nil && l.HandshakeLog.ServerHello != nil
+	}
+	return false
+}
+
 // Scan TODO: describe how scan operates in appropriate detail
 //1. Send a request (currently get-printer-attributes)
 //2. Take in that response & read out version numbers
 func (scanner *Scanner) Scan(target zgrab2.ScanTarget) (zgrab2.ScanStatus, interface{}, error) {
-	scan := scanner.newIPPScan(&target)
-	//defer scan.Cleanup()
-	var err *zgrab2.ScanError
-	// Try all known IPP versions from newest to oldest until version is supported
-	for i := 0; i < len(Versions); i++ {
-		err = scanner.Grab(scan, &target, &Versions[i])
-		if err == nil || (err != nil && err.Err != ErrVersionNotSupported) {
-			break
-		}
-		if i == len(Versions) - 1 && err.Err == ErrVersionNotSupported {
-			return zgrab2.SCAN_APPLICATION_ERROR, &scan.results, err.Err
-		}
-	}
+	// Try all known IPP versions from newest to oldest until we reach a supported version
+	scan, err := scanner.tryGrabForVersions(&target, &Versions)
 	if err != nil {
-		// Adapted from http module's RetryHTTPS logic
+		// If versionNotSupported error was confirmed, the scanner was connecting w/o TLS, so don't retry
+		// Same goes for a protocol error of any kind. It means we got something back but it didn't conform.
+		if err.Status == zgrab2.SCAN_APPLICATION_ERROR || err.Status == zgrab2.SCAN_PROTOCOL_ERROR {
+			return err.Unpack(&scan.results)
```

Comment:
It looks like this will always return a non-nil result (so always indicating a detection in these cases) -- is that the intent? Since `scan.results` is (IIRC) a `ScanResults` and not a `*ScanResults`, its default/not-yet-set value will be `{}`, not `nil`.

Reply:
Oops sorry for not responding. My intention is to return the partial result despite the error in this case, because each of these statuses indicate a positive identification as IPP but with an error.